rpc/filter_api.go

```diff
@@ -1,79 +1,555 @@
+	hashes   []common.Hash
+	crit     filters.FilterCriteria
+	logs     []*ethtypes.Log
+	s        *Subscription // associated subscription in event system
+}
+
+// PublicFilterAPI offers support to create and manage filters. This will allow external clients to retrieve various
+// information related to the Ethereum protocol such as blocks, transactions and logs.
 type PublicFilterAPI struct {
-	cliCtx  context.CLIContext
-	backend Backend
-	filters map[rpc.ID]*Filter // ID to filter; TODO: change to sync.Map in case of concurrent writes
+	cliCtx    clientcontext.CLIContext
+	backend   FiltersBackend
+	events    *EventSystem
+	filtersMu sync.Mutex
+	filters   map[rpc.ID]*filter
 }
 
-// NewPublicEthAPI creates an instance of the public ETH Web3 API.
-func NewPublicFilterAPI(cliCtx context.CLIContext, backend Backend) *PublicFilterAPI {
-	return &PublicFilterAPI{
+// NewPublicFilterAPI returns a new PublicFilterAPI instance.
+func NewPublicFilterAPI(cliCtx clientcontext.CLIContext, backend FiltersBackend) *PublicFilterAPI {
+	// start the client to subscribe to Tendermint events
+	err := cliCtx.Client.Start()
+	if err != nil {
+		panic(err)
+	}
+
+	api := &PublicFilterAPI{
 		cliCtx:  cliCtx,
 		backend: backend,
-		filters: make(map[rpc.ID]*Filter),
+		filters: make(map[rpc.ID]*filter),
+		events:  NewEventSystem(cliCtx.Client),
 	}
+
+	go api.timeoutLoop()
+
+	return api
 }
 
-// NewFilter instantiates a new filter.
-func (e *PublicFilterAPI) NewFilter(criteria filters.FilterCriteria) rpc.ID {
-	id := rpc.NewID()
-	e.filters[id] = NewFilter(e.backend, &criteria)
-	return id
+// timeoutLoop runs every 5 minutes and deletes filters that have not been recently used.
+// Tt is started when the api is created.
+func (api *PublicFilterAPI) timeoutLoop() {
+	ticker := time.NewTicker(deadline)
+	defer ticker.Stop()
+
+	for {
+		<-ticker.C
+		api.filtersMu.Lock()
+		for id, f := range api.filters {
+			select {
+			case <-f.deadline.C:
+				f.s.Unsubscribe(api.events)
+				delete(api.filters, id)
+			default:
+				continue
+			}
+		}
+		api.filtersMu.Unlock()
+	}
 }
 
-// NewBlockFilter instantiates a new block filter.
-func (e *PublicFilterAPI) NewBlockFilter() rpc.ID {
-	id := rpc.NewID()
-	e.filters[id] = NewBlockFilter(e.backend)
-	return id
+// NewPendingTransactionFilter creates a filter that fetches pending transaction hashes
+// as transactions enter the pending state.
+//
+// It is part of the filter package because this filter can be used through the
+// `eth_getFilterChanges` polling method that is also used for log filters.
+//
+// https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_newPendingTransactionFilter
+func (api *PublicFilterAPI) NewPendingTransactionFilter() rpc.ID {
+	pendingTxSub, cancelSubs, err := api.events.SubscribePendingTxs()
+	if err != nil {
+		// wrap error on the ID
+		return rpc.ID(fmt.Sprintf("error creating pending tx filter: %s", err.Error()))
+	}
+
+	defer cancelSubs()
```

Comment:
i think this should be put into the goroutine, since it will call `cancelSubs` to be called once the function returns, but the goroutine should still be running (assuming that cancelSubs does what i think it does)